Consolidate VMINMON and cleanup monitor scripts

KateFriedman-NOAA opened the pull request:
# Description

This PR combines the GDAS and GFS versions of the VMINMON JJOB and ex-driver scripts into single GLOBAL/global versions. Additionally:

* some further cleanup of the GSI-Monitor scripts is done; variable settings in ex-driver scripts are moved up to the JJOB level
* the fix symlinks are consolidated into a new `fix/mon` subfolder (formerly `fix/gdas`) and copies of the gfs monitor fix files under `fix/products` are removed

Resolves #1925

# Type of change

- Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?

- Short low-res atmos-only cycled test on WCOSS2

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] I have made corresponding changes to the documentation if necessary

KateFriedman-NOAA committed:
Rename and delete VMINMON scripts

- Rename JGDAS_ATMOS_VMINMON -> JGLOBAL_ATMOS_VMINMON
and exgdas_atmos_vminmon.sh -> exglobal_atmos_vminmon.sh
- Delete JGFS_ATMOS_VMINMON and exgfs_atmos_vminmon.sh
- Will consolidate around GLOBAL versions

Refs #1925

KateFriedman-NOAA committed:
Monitor fix file updates

- Delete gfs_minmon_*.txt fix files from fix/product.
- Update link script to link to gfs_minmon_*.txt files
from GSI-Monitor package checkout under /sorc.
- Rename generated fix folder "gdas" to "mon" and place
gfs fix file symlinks alongside gdas ones in folder.
- Update .gitignore for new fix/mon folder.

Refs #1925

KateFriedman-NOAA committed:
Consolidate new GLOBAL VMINMON scripts

- Update the new GLOBAL/global versions of the vminmon
scripts to consolidate gdas and gfs functionality.
- Clean up scripts some.
- Update rocoto job vminmon.sh script to use consolidated
JJOB script now.

Refs #1925

KateFriedman-NOAA committed:
Export gsistat for ex-script

Add an "export" for the gsistat variable in JGLOBAL_ATMOS_VMINMON
so the ex-script gets it from the environment.

Refs #1925

KateFriedman-NOAA committed:
Update fix subfolder to "mon"

Also update the fix subfolder name in the JJOB scripts
for the oznmon and radmon jobs.

Refs #1925

KateFriedman-NOAA committed:
Clean up the GSI monitor scripts some more

Refs #1925

aerorahul commented on scripts/exglobal_atmos_vminmon.sh:
```diff
@@ -19,39 +19,11 @@ source "$HOMEgfs/ush/preamble.sh"
-   mkdir $DATA
-fi
-cd $DATA
+export PDY=${1:-${PDY:?}}
+export cyc=${2:-${cyc:?}}
```
This will be set in the j-job already.  
We don't do this in any exscript, so why add here?

KateFriedman-NOAA replied:
I pulled those two lines in from `exgfs_atmos_vminmon.sh` during the consolidation but after further review we don't need to check these arguments in the ex-script or even send them in as arguments in the JJOB script since they should already be in the environment from the workflow manager.

Have removed these now @ 2e76dbc8

aerorahul commented on jobs/JGLOBAL_ATMOS_VMINMON:
```diff
@@ -1,42 +1,25 @@
-export EXECminmon=${EXECminmon:-${HOMEminmon}/exec}
-export USHminmon=${USHminmon:-${HOMEminmon}/ush}
-
+export USHminmon=${USHminmon:-${HOMEgfs}/ush}
```
Shoulnd't this be `USHgfs`?

KateFriedman-NOAA replied:
Have changed it to `USHgfs`.

aerorahul commented on jobs/JGLOBAL_ATMOS_VMINMON:
```diff
@@ -1,42 +1,25 @@
 
 #############################################
-#  determine PDY and cyc for previous cycle
+# Determine PDY and cyc for previous cycle
```
The convention is `GDATE` for previous cycle.  Is there a reason why GSI monitoring is defining its own names?  
Should use `${assim_freq}` instead of `-6` below.

KateFriedman-NOAA replied:
Have updated to use `GDATE`, `gPDY`, `gcyc`, and `assim_freq`.

aerorahul commented on jobs/JGDAS_ATMOS_VERFRAD:
```diff
@@ -6,40 +6,34 @@
-
 export RADMON_SUFFIX=${RADMON_SUFFIX:-${RUN}}
 export CYCLE_INTERVAL=${CYCLE_INTERVAL:-6}
+export VERBOSE=${VERBOSE:-YES}
```
`CYCLE_INTERVA` is the same as `assim_freq`

KateFriedman-NOAA replied:
Since I see `CYCLE_INTERVAL` within scripts in the GSI-Monitor package I have left it being set in the script but have changed it to use `assim_freq`.

aerorahul reviewed (commented):
For a while the GSI monitoring scripts were redefining its own variable names not consistent with the GFS.
Do we (or @EdwardSafford-NOAA) have any plans on making them consistent or will they live in parallel universes for the foreseeable future?

KateFriedman-NOAA committed:
Resolve SC2086 shellcheck warnings for VMINMON

Refs #1925

KateFriedman-NOAA committed:
Remove PDY and cyc arguments for VMINMON scripts

Refs #1925

aerorahul commented on jobs/JGDAS_ATMOS_VERFRAD:
```diff
@@ -57,31 +51,26 @@ export TANKverf_radM1=${TANKverf_radM1:-${COM_ATMOS_RADMON_PREV}}
 export radstat=${radstat:-${COM_ATMOS_ANALYSIS}/gdas.t${cyc}z.radstat}
+export satype_file=${satype_file:-${FIXgdas}/gdas_radmon_satype.txt}
 
-echo " "
-echo "JOB HAS STARTED"
-echo " "
-
+#  Other variables
+export RAD_AREA=${RAD_AREA:-glb}
+export MAKE_CTL=${MAKE_CTL:-1}
+export MAKE_DATA=${MAKE_DATA:-1}
+export USE_ANL=${USE_ANL:-1}
+export PDATE=${PDY}${cyc}
+export DO_DIAG_RPT=${DO_DIAG_RPT:-1}
+export DO_DATA_RPT=${DO_DATA_RPT:-1}
```
Do we really need pass-through capability here?
For e.g. `RAD_AREA="glb"` should be sufficient.  Some of these could be even elevated to the configs.
`PDATE` can be exported at line 36 replacing `pdate`.  Using 2 variables with the same name one uppercase and another lowercase should be unnecessary.

KateFriedman-NOAA replied:
Moving settings to configs. The uppercase `PDATE` is current cycle, while the lower case `pdate` was the prior cycle but is now being changed to `GDATE`.

KateFriedman-NOAA committed:
Address reviewer comments with changes

Refs #1925

WalterKolczynski-NOAA commented on .gitignore:
```diff
@@ -30,11 +30,11 @@ fix/chem
 fix/gsi
 fix/lut
 fix/mom6
+fix/mon
```
Can we make this "monitor", maybe even "da_monitor"? I want to get in the habit of using descriptive names instead of short inscrutable ones.

WalterKolczynski-NOAA replied:
Also see below about whether this should be `fix` at all instead of `parm`.

KateFriedman-NOAA replied:
Yeah...I am moving these "fix" files to `parm/monitor` (which is being renamed from `parm/mon`. Will discontinue use of `fix/mon`.

WalterKolczynski-NOAA commented on jobs/JGDAS_ATMOS_VERFOZN:
```diff
@@ -15,29 +15,19 @@ export HOMEgfs_ozn=${HOMEgfs:-${NWROOT}/gfs.${gfs_ver}}
-
-export P_PDY=${pdate:0:8}
-export p_cyc=${pdate:8:2}
+GDATE=$(${NDATE} -"${assim_freq}" "${PDY}${cyc}")
```
Rahul has been gradually changing these to use *nix `date` instead of NDATE:
```suggestion
GDATE=$(date --utc +%Y%m%d%H -d "${PDY} ${cyc} - ${assim_freq} hours")
```

WalterKolczynski-NOAA commented on jobs/JGDAS_ATMOS_VERFRAD:
```diff
@@ -6,82 +6,71 @@
-
-export P_PDY=${pdate:0:8}
-export p_cyc=${pdate:8:2}
+GDATE=$(${NDATE} -"${assim_freq}" "${PDY}${cyc}")
```
Ditto.

KateFriedman-NOAA replied:
Will update as well, thanks!

WalterKolczynski-NOAA commented on jobs/JGDAS_ATMOS_VERFOZN:
```diff
@@ -49,20 +39,27 @@ export TANKverf_ozn=${TANKverf_ozn:-${COM_ATMOS_OZNMON}}
+#-----------------------------------
+# Other variables
+#
+export OZN_AREA=${OZN_AREA:-glb}
+export oznstat=${oznstat:-${COM_ATMOS_ANALYSIS}/gdas.t${cyc}z.oznstat}
+export SATYPE_FILE=${SATYPE_FILE:-${FIXgdas_ozn}/gdas_oznmon_satype.txt}
+export PDATE=${PDY}${cyc}
+export DO_DATA_RPT=${DO_DATA_RPT:-1}
```
Except `PDATE`, these should probably be moved to a config file.

KateFriedman-NOAA replied:
Moving.

WalterKolczynski-NOAA commented on jobs/JGDAS_ATMOS_VERFRAD:
```diff
@@ -6,82 +6,71 @@
-echo "JOB HAS STARTED"
-echo " "
+#  Other variables
+export RAD_AREA=${RAD_AREA:-glb}
+export MAKE_CTL=${MAKE_CTL:-1}
+export MAKE_DATA=${MAKE_DATA:-1}
+export USE_ANL=${USE_ANL:-1}
+export PDATE=${PDY}${cyc}
+export DO_DIAG_RPT=${DO_DIAG_RPT:-1}
+export DO_DATA_RPT=${DO_DATA_RPT:-1}
```
Except `PDATE`, these should be probably moved to a config file.

KateFriedman-NOAA replied:
Moving.

WalterKolczynski-NOAA commented on jobs/JGDAS_ATMOS_VERFRAD:
```diff
@@ -6,82 +6,71 @@
-   echo "JOB HAS COMPLETED NORMALLY"
-   echo " "
-fi
+[[ ${err} -ne 0 ]] && exit ${err}
```
I'm not a fan of using short-circuiting in place of if-then
```suggestion
if (( err != 0 )); then
    exit "${err}"
fi
```

Or, if someone insists on the short-circuit:
```suggestion
(( err != 0 )) && exit "${err}"
```

WalterKolczynski-NOAA commented on jobs/JGLOBAL_ATMOS_VMINMON:
```diff
@@ -1,66 +1,54 @@
-export P_PDY=${pdate:0:8}
-export p_cyc=${pdate:8:2}
-
+GDATE=$(${NDATE} -"${assim_freq}" "${PDY}${cyc}")
```
As before.

KateFriedman-NOAA replied:
Also updating

WalterKolczynski-NOAA commented on jobs/JGLOBAL_ATMOS_VMINMON:
```diff
@@ -1,66 +1,54 @@
+########################################
+#  Filenames
+########################################
+export gsistat=${gsistat:-${COM_ATMOS_ANALYSIS}/${RUN}.t${cyc}z.gsistat}
+export mm_gnormfile=${gnormfile:-${M_FIX}/${RUN}_minmon_gnorm.txt}
+export mm_costfile=${costfile:-${M_FIX}/${RUN}_minmon_cost.txt}
```
Config file?

KateFriedman-NOAA replied:
Moving to config

WalterKolczynski-NOAA commented on sorc/link_workflow.sh:
```diff
@@ -215,10 +215,12 @@ fi
+  cd mon || exit 1
   ${LINK_OR_COPY} "${HOMEgfs}/sorc/gsi_monitor.fd/src/Minimization_Monitor/nwprod/gdas/fix/gdas_minmon_cost.txt" .
   ${LINK_OR_COPY} "${HOMEgfs}/sorc/gsi_monitor.fd/src/Minimization_Monitor/nwprod/gdas/fix/gdas_minmon_gnorm.txt" .
+  ${LINK_OR_COPY} "${HOMEgfs}/sorc/gsi_monitor.fd/src/Minimization_Monitor/nwprod/gfs/fix/gfs_minmon_cost.txt" .
+  ${LINK_OR_COPY} "${HOMEgfs}/sorc/gsi_monitor.fd/src/Minimization_Monitor/nwprod/gfs/fix/gfs_minmon_gnorm.txt" .
```
These are just text files. Shouldn't they be in `parm/`?

KateFriedman-NOAA replied:
Moving to `parm/monitor`.

KateFriedman-NOAA committed:
Update GDATE setting in JGDAS_ATMOS_VERFOZN

Refs #1925

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

KateFriedman-NOAA committed:
Update exit err catch in JGDAS_ATMOS_VERFRAD

Refs #1925

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

KateFriedman-NOAA committed:
Move fix to parm

- Change symlinks for monitor "fix" files to go to parm.
- Rename "parm/mon" to "parm/monitor".

Refs #1925

KateFriedman-NOAA committed:
Further cleanup of monitor scripts

- Address reviewer comments and further clean up the monitor scripts.
- Remove use of PDATE and replace with PDY and cyc variables.
- Move settings from JJOB scripts to configs.

Refs #1925

WalterKolczynski-NOAA commented on ush/radmon_verf_time.sh:
```diff
@@ -137,17 +126,16 @@ fi
-ihh=$(echo $PDATE | cut -c9-10)
-cyc=$ihh
-CYCLE=$cyc
+iyy=$(echo $PDY | cut -c1-4)
+imm=$(echo $PDY | cut -c5-6)
+idd=$(echo $PDY | cut -c7-8)
```
```suggestion
iyy="${PDY:0:4}"
imm="${PDY:4:2}"
idd="${PDY:6:2}"
```

WalterKolczynski-NOAA commented on ush/radmon_verf_bcor.sh:
```diff
@@ -111,10 +104,10 @@ else
-   imm=$(echo $PDATE | cut -c5-6)
-   idd=$(echo $PDATE | cut -c7-8)
-   ihh=$(echo $PDATE | cut -c9-10)
+   iyy=$(echo ${PDY} | cut -c1-4)
+   imm=$(echo ${PDY} | cut -c5-6)
+   idd=$(echo ${PDY} | cut -c7-8)
```
```suggestion
   iyy="${PDY:0:4}"
   imm="${PDY:4:2}"
   idd="${PDY:6:2}"
```

WalterKolczynski-NOAA commented on ush/radmon_verf_bcoef.sh:
```diff
@@ -118,10 +107,10 @@ else
-   imm=$(echo $PDATE | cut -c5-6)
-   idd=$(echo $PDATE | cut -c7-8)
-   ihh=$(echo $PDATE | cut -c9-10)
+   iyy=$(echo ${PDY} | cut -c1-4)
+   imm=$(echo ${PDY} | cut -c5-6)
+   idd=$(echo ${PDY} | cut -c7-8)
```
```suggestion
   iyy="${PDY:0:4}"
   imm="${PDY:4:2}"
   idd="${PDY:6:2}"
```

KateFriedman-NOAA replied:
Making change, thanks!

WalterKolczynski-NOAA commented on parm/config/gfs/config.verfozn:
```diff
@@ -6,4 +6,19 @@ echo "BEGIN: config.verfozn"
+export SATYPE_FILE=${PARMmonitor}/gdas_oznmon_satype.txt
+
+# Source the parm file
+export parm_file=${PARMmonitor}/gdas_oznmon.parm
+. "${parm_file}"
```
Is it necessary to both source the file and export the path to the file? And is the file coming from elsewhere, or should those settings also be added here?

KateFriedman-NOAA replied:
I can compress those two lines. That parm file comes from the GSI-Monitor package under `/sorc` and gets symlinked during setup. It has settings we could pull into the config and then ignore the parm file from the package:
```
kate.friedman@clogin09:/lfs/h2/emc/global/noscrub/kate.friedman/git/feature-combine_vminmon_scripts> cat parm/monitor/gdas_oznmon.parm
##################################
# specify oznmon utilities
##################################
export COMPRESS=${COMPRESS:-gzip}
export UNCOMPRESS=${UNCOMPRESS:-gunzip}
export VALIDATE_DATA=${VALIDATE_DATA:-1}
export Z=${Z:-gz}
kate.friedman@clogin09:/lfs/h2/emc/global/noscrub/kate.friedman/git/feature-combine_vminmon_scripts> ll parm/monitor/gdas_oznmon.parm
lrwxrwxrwx 1 kate.friedman global 155 Nov 28 13:42 parm/monitor/gdas_oznmon.parm -> /lfs/h2/emc/global/noscrub/kate.friedman/git/feature-combine_vminmon_scripts/sorc/gsi_m
onitor.fd/src/Ozone_Monitor/nwprod/gdas_oznmon/parm/gdas_oznmon.parm
```

WalterKolczynski-NOAA commented on parm/config/gfs/config.verfrad:
```diff
@@ -6,4 +6,23 @@ echo "BEGIN: config.verfrad"
+
+# Source the parm file
+export parm_file=${PARMmonitor}/da_mon.parm
+. "${parm_file}"
```
Same question.

KateFriedman-NOAA replied:
Similar thing:
```
kate.friedman@clogin09:/lfs/h2/emc/global/noscrub/kate.friedman/git/feature-combine_vminmon_scripts> ll parm/monitor/da_mon.parm
lrwxrwxrwx 1 kate.friedman global 158 Nov 28 13:42 parm/monitor/da_mon.parm -> /lfs/h2/emc/global/noscrub/kate.friedman/git/feature-combine_vminmon_scripts/sorc/gsi_monito
r.fd/src/Radiance_Monitor/nwprod/gdas_radmon/parm/gdas_radmon.parm
kate.friedman@clogin09:/lfs/h2/emc/global/noscrub/kate.friedman/git/feature-combine_vminmon_scripts> cat parm/monitor/da_mon.parm
##################################
# specify radmon utilities
##################################
export LITTLE_ENDIAN=${LITTLE_ENDIAN:-0}
export COMPRESS=${COMPRESS:-gzip}
export UNCOMPRESS=${UNCOMPRESS:-gunzip}
export Z=${Z:-gz}
```

WalterKolczynski-NOAA commented on ush/ozn_xtrct.sh:
```diff
@@ -149,14 +149,14 @@ else
-      imm=$(echo ${PDATE} | cut -c5-6)
-      idd=$(echo ${PDATE} | cut -c7-8)
-      ihh=$(echo ${PDATE} | cut -c9-10)
+      iyy=$(echo ${PDY} | cut -c1-4)
+      imm=$(echo ${PDY} | cut -c5-6)
+      idd=$(echo ${PDY} | cut -c7-8)
```
```suggestion
      iyy="${PDY:0:4}"
      imm="${PDY:4:2}"
      idd="${PDY:6:2}"
```

KateFriedman-NOAA replied:
Making change, thanks!

WalterKolczynski-NOAA commented on ush/radmon_verf_angle.sh:
```diff
@@ -125,10 +118,10 @@ else
-   imm=$(echo $PDATE | cut -c5-6)
-   idd=$(echo $PDATE | cut -c7-8)
-   ihh=$(echo $PDATE | cut -c9-10)
+   iyy=$(echo ${PDY} | cut -c1-4)
+   imm=$(echo ${PDY} | cut -c5-6)
+   idd=$(echo ${PDY} | cut -c7-8)
```
```suggestion
   iyy="${PDY:0:4}"
   imm="${PDY:4:2}"
   idd="${PDY:6:2}"
```

WalterKolczynski-NOAA reviewed (changes requested):
Also needs a healthy sprinkling of double-quotes to make shellcheck happy.

KateFriedman-NOAA committed:
Address shellcheck warnings

- Resolve some SC2086 and SC2250 warnings.

Refs #1925

KateFriedman-NOAA committed:
Update PDY parsing in ush/radmon_verf_angle.sh

Refs #1925

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

KateFriedman-NOAA committed:
Update PDY parsing in ush/radmon_verf_bcor.sh

Refs #1925

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/combine_vminmon_scripts

KateFriedman-NOAA committed:
Compress sourcing of parm file in configs

Refs #1925

KateFriedman-NOAA committed:
Address SC2086 and SC2250 shellcheck warnings

- In monitor ush scripts

Refs #1925

KateFriedman-NOAA committed:
Address SC2086 shellcheck warnings in exgdas_atmos_verfrad.sh

Refs #1925

KateFriedman-NOAA committed:
Cleanup ozn_xtrct.sh a bit

Refs #1925

KateFriedman-NOAA committed:
Remove invocation of script that no longer exists

The ush/clean_tankdir.sh script no longer exists in either
global-workflow or GSI-Monitor. Remove it's invocation.

Refs #1925

KateFriedman-NOAA committed:
Address SC2086 in exglobal_atmos_vminmon.sh

Refs #1925

KateFriedman-NOAA committed:
A few corrections to radmon scripts

Refs #1925

aerorahul commented on jobs/JGDAS_ATMOS_VERFOZN:
```diff
@@ -6,69 +6,32 @@
-
-export P_PDY=${pdate:0:8}
-export p_cyc=${pdate:8:2}
+GDATE=$(date --utc +%Y%m%d%H -d "${PDY} ${cyc} - ${assim_freq} hours")
```
```suggestion
# shellcheck disable=SC2153
GDATE=$(date --utc +%Y%m%d%H -d "${PDY} ${cyc} - ${assim_freq} hours")
```

aerorahul commented on scripts/exgdas_atmos_verfrad.sh:
```diff
@@ -66,11 +29,11 @@ if [[ -s ${radstat} && -s ${biascr} ]]; then
 
-   $NCP $biascr  ./biascr.$PDATE
-   $NCP $radstat ./radstat.$PDATE
+   $NCP "$biascr"  "./biascr.${PDY}${cyc}"
+   $NCP "$radstat" "./radstat.${PDY}${cyc}"
```
```suggestion
   $NCP "${biascr}"  "./biascr.${PDY}${cyc}"
   $NCP "${radstat}" "./radstat.${PDY}${cyc}"
```

KateFriedman-NOAA committed:
Add {} into scripts/exgdas_atmos_verfrad.sh

Refs #1925

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

KateFriedman-NOAA committed:
Add shellcheck disable to jobs/JGDAS_ATMOS_VERFOZN

Refs #1925

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

KateFriedman-NOAA committed:
Add shellcheck disable for SC2153 for monitor jobs

Disable shellcheck SC2153 for PDY in GDATE

Refs #1925

KateFriedman-NOAA commented:
The error in `/work2/noaa/stmp/GFS_CI_ROOT/PR/2088/RUNTESTS/COMROT/C48_ATM_d280a02d/logs/2021032312/gfsmetpg2g1.log`:
```
+ metp.sh[13]: /work2/noaa/stmp/GFS_CI_ROOT/PR/2088/global-workflow/jobs/JGFS_ATMOS_VERIFICATION
/work2/noaa/stmp/GFS_CI_ROOT/PR/2088/global-workflow/jobs/rocoto/metp.sh: line 13: /work2/noaa/stmp/GFS_CI_ROOT/PR/2088/global-workflow/jobs/JGFS_ATMOS_VERIFICATION: Permission denied
+ metp.sh[1]: postamble metp.sh 1701273324 1
+ preamble.sh[68]: set +x
End metp.sh at 15:55:32 with error code 1 (time elapsed: 00:00:08)
```
Permissions need to be opened up for the new `JGFS_ATMOS_VERIFICATION`:
```
-rwxr-x--- 1 mterry stmp 2021 Nov 29 07:52 JGFS_ATMOS_POSTSND*
-rw-r----- 1 mterry stmp 1712 Nov 29 07:52 JGFS_ATMOS_VERIFICATION
-rwxr-x--- 1 mterry stmp 1653 Nov 29 07:52 JGLOBAL_AERO_ANALYSIS_FINALIZE*
```
Looks like the permissions in `develop` need to be opened up, I see the same thing with a fresh clone of `develop` on Orion and Hera.

@TerrenceMcGuinness-NOAA can you adjust the permissions in the CI clones on Orion and Hera and resubmit the failed metp jobs for now? Thanks!

TerrenceMcGuinness-NOAA commented:
Oh yes I just saw this come up too, I think it is an issue with ${USER} not getting through the YAML reader, looking into it now.  That path is supposed to be `/work2/noaa/stmp/${USER}/GFS_CI_ROOT/PR/2088/global-workflow/jobs/JGFS_ATMOS_VERIFICATION`

OK, the path is correct, still  looking into the root cause ...

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/combine_vminmon_scripts

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/combine_vminmon_scripts

WalterKolczynski-NOAA commented on scripts/exgdas_atmos_verfrad.sh:
```diff
@@ -82,10 +45,10 @@ if [[ -s ${radstat} && -s ${biascr} ]]; then
    fi
 
-   echo satype_file = $satype_file
+   echo satype_file = "$satype_file"
```
```suggestion
   echo satype_file = "${satype_file}"
```

WalterKolczynski-NOAA commented on scripts/exgdas_atmos_verfrad.sh:
```diff
@@ -82,10 +45,10 @@ if [[ -s ${radstat} && -s ${biascr} ]]; then
    radstat_satype=$(ls d*ges* | awk -F_ '{ print $2 "_" $3 }')
    if [[ "$VERBOSE" = "YES" ]]; then
-      echo $radstat_satype
+      echo "$radstat_satype"
```
```suggestion
      echo "${radstat_satype}"
```

WalterKolczynski-NOAA commented on scripts/exgdas_atmos_verfrad.sh:
```diff
@@ -98,8 +61,8 @@ if [[ -s ${radstat} && -s ${biascr} ]]; then
 
-   echo satype_file = $satype_file
-   export SATYPE=$(cat ${satype_file})
+   echo satype_file = "$satype_file"
```
```suggestion
   echo satype_file = "${satype_file}"
```

WalterKolczynski-NOAA commented on scripts/exgdas_atmos_verfrad.sh:
```diff
@@ -110,7 +73,7 @@ if [[ -s ${radstat} && -s ${biascr} ]]; then
-      test=$(echo $SATYPE | grep $type | wc -l)
+      test=$(echo "$SATYPE" | grep "$type" | wc -l)
 
       if [[ $test -eq 0 ]]; then
```
`test` is a bash command so we should probably avoid using it as a variable name
```suggestion
      type_count=$(echo "${SATYPE}" | grep "${type}" | wc -l)

      if (( type_count == 0 )); then
```

WalterKolczynski-NOAA commented on scripts/exgdas_atmos_verfrad.sh:
```diff
@@ -129,23 +92,23 @@ if [[ -s ${radstat} && -s ${biascr} ]]; then
    for type in ${SATYPE}; do
 
-      if [[ netcdf -eq 0 && -e diag_${type}_ges.${PDATE}.nc4.${Z} ]]; then
+      if [[ netcdf -eq 0 && -e diag_${type}_ges.${PDY}${cyc}.nc4.${Z} ]]; then
```
As written, `netcdf` is a literal and not the variable, so it will never equal zero.
```suggestion
      if (( netcdf == 0 )) && [[ -e "diag_${type}_ges.${PDY}${cyc}.nc4.${Z}" ]]; then
```

WalterKolczynski-NOAA commented on scripts/exgdas_atmos_verfrad.sh:
```diff
@@ -203,7 +166,7 @@ export CHGRP_CMD=${CHGRP_CMD:-"chgrp ${group_name:-rstprod}"}
 for rtype in $rlist; do
   if compgen -G "$TANKverf_rad/*${rtype}*" > /dev/null; then
-     ${CHGRP_CMD} "${TANKverf_rad}"/*${rtype}*
+     ${CHGRP_CMD} "${TANKverf_rad}/*${rtype}*"
```
Wildcards won't expand within quotes.
```suggestion
     ${CHGRP_CMD} "${TANKverf_rad}/"*"${rtype}"*
```

WalterKolczynski-NOAA commented on ush/radmon_diag_ck.sh:
```diff
@@ -71,23 +71,23 @@ echo "--> radmon_diag_ck.sh"
    #  load contents of satype_file into an array
    #
-   satype_contents=`cat ${satype_file}`
+   satype_contents=`cat "${satype_file}"`
```
Kill the evil backticks.
```suggestion
   satype_contents=$(cat "${satype_file}")
```

WalterKolczynski-NOAA commented on ush/radmon_diag_ck.sh:
```diff
@@ -71,23 +71,23 @@ echo "--> radmon_diag_ck.sh"
+   for sat in ${satype_contents}; do
+     test=`echo "${radstat_contents}" | grep "${sat}"`
      
      if [[ ${#test} -le 0 ]]; then
```
For consistency with previous.
```suggestion
     content_count=$(echo "${radstat_contents}" | grep "${sat}" | wc -l)

     if (( content_count <= 0 )); then
```

WalterKolczynski-NOAA commented on ush/radmon_err_rpt.sh:
```diff
@@ -63,28 +63,28 @@ err=0
 
 have_diag_rpt=0
-if [[ -s $diag_rpt ]]; then
+if [[ -s ${diag_rpt} ]]; then
```
```suggestion
if [[ -s "${diag_rpt}" ]]; then
```

WalterKolczynski-NOAA commented on ush/radmon_err_rpt.sh:
```diff
@@ -96,95 +96,95 @@ echo "have_diag_rpt = $have_diag_rpt"
          match_len=$(echo ${#match})
-         if [[ $match_len > 0 ]]; then
-            channel2=$(echo $match | gawk '{print $3}')
+         if [[ ${match_len} > 0 ]]; then
```
As-is, this doesn't do what was intended.
```suggestion
         if (( match_len > 0 )); then
```

WalterKolczynski-NOAA commented on ush/radmon_err_rpt.sh:
```diff
@@ -96,95 +96,95 @@ echo "have_diag_rpt = $have_diag_rpt"
       match_len=$(echo ${#match})
          
-      if [[ $match_len > 0 ]]; then
+      if [[ ${match_len} > 0 ]]; then
```
```suggestion
      if (( match_len > 0 )); then
```

WalterKolczynski-NOAA commented on ush/radmon_verf_angle.sh:
```diff
@@ -100,22 +93,22 @@ LITTLE_ENDIAN=${LITTLE_ENDIAN:-0}
 err=0
 angle_exec=radmon_angle.x
-shared_scaninfo=${shared_scaninfo:-$FIXgdas/gdas_radmon_scaninfo.txt}
+shared_scaninfo=${shared_scaninfo:-${PARMmonitor}/gdas_radmon_scaninfo.txt}
```
```suggestion
shared_scaninfo="${shared_scaninfo:-${PARMmonitor}/gdas_radmon_scaninfo.txt}"
```

WalterKolczynski-NOAA commented on ush/radmon_verf_angle.sh:
```diff
@@ -143,24 +136,24 @@ else
          prep_step
 
-         ctr=$(expr $ctr + 1)
+         ctr=$(expr ${ctr} + 1)
```
```suggestion
         ctr=$((ctr + 1))
```

WalterKolczynski-NOAA commented on ush/radmon_verf_angle.sh:
```diff
@@ -143,24 +136,24 @@ else
-         if [[ $dtype == "anl" ]]; then
-            data_file=${type}_anl.${PDATE}.ieee_d
+         if [[ ${dtype} == "anl" ]]; then
+            data_file=${type}_anl.${PDY}${cyc}.ieee_d
```
```suggestion
            data_file="${type}_anl.${PDY}${cyc}.ieee_d"
```

WalterKolczynski-NOAA commented on ush/radmon_verf_angle.sh:
```diff
@@ -143,24 +136,24 @@ else
             angl_ctl=angle.${ctl_file}
          else
-            data_file=${type}.${PDATE}.ieee_d
+            data_file=${type}.${PDY}${cyc}.ieee_d
```
```suggestion
            data_file="${type}.${PDY}${cyc}.ieee_d"
```

WalterKolczynski-NOAA commented on ush/radmon_verf_angle.sh:
```diff
@@ -187,18 +180,18 @@ cat << EOF > input
 
 	 startmsg
-         ./${angle_exec} < input >>   ${pgmout} 2>>errfile
+         ./${angle_exec} < input >>   "${pgmout}" 2>>errfile
```
```suggestion
         ./${angle_exec} < input >> "${pgmout}" 2>>errfile
```

WalterKolczynski-NOAA commented on ush/radmon_verf_angle.sh:
```diff
@@ -187,18 +180,18 @@ cat << EOF > input
-         if [[ $err -ne 0 ]]; then
-             fail=$(expr $fail + 1)
+         if [[ ${err} -ne 0 ]]; then
+             fail=$(expr ${fail} + 1)
```
```suggestion
             fail=$(( fail + 1 ))
```

WalterKolczynski-NOAA commented on ush/radmon_verf_angle.sh:
```diff
@@ -207,24 +200,24 @@ EOF
    tar_file=radmon_angle.tar
    if compgen -G "angle*.ieee_d*" > /dev/null || compgen -G "angle*.ctl*" > /dev/null; then
-      tar -cf $tar_file angle*.ieee_d* angle*.ctl*
+      tar -cf ${tar_file} angle*.ieee_d* angle*.ctl*
```
```suggestion
      tar -cf "${tar_file}" angle*.ieee_d* angle*.ctl*
```

WalterKolczynski-NOAA commented on ush/radmon_verf_bcoef.sh:
```diff
@@ -140,19 +129,19 @@ else
          prep_step
 
-         ctr=$(expr $ctr + 1)
+         ctr=$(expr ${ctr} + 1)
```
```suggestion
         ctr=$(( ctr + 1 ))
```

WalterKolczynski-NOAA commented on ush/radmon_verf_bcoef.sh:
```diff
@@ -140,19 +129,19 @@ else
-         if [[ $dtype == "anl" ]]; then
-            data_file=${type}_anl.${PDATE}.ieee_d
+         if [[ ${dtype} == "anl" ]]; then
+            data_file=${type}_anl.${PDY}${cyc}.ieee_d
```
```suggestion
            data_file="${type}_anl.${PDY}${cyc}.ieee_d"
```

WalterKolczynski-NOAA commented on ush/radmon_verf_bcoef.sh:
```diff
@@ -140,19 +129,19 @@ else
             bcoef_ctl=bcoef.${ctl_file}
          else
-            data_file=${type}.${PDATE}.ieee_d
+            data_file=${type}.${PDY}${cyc}.ieee_d
```
```suggestion
            data_file="${type}.${PDY}${cyc}.ieee_d"
```

WalterKolczynski-NOAA commented on ush/radmon_verf_bcoef.sh:
```diff
@@ -180,10 +169,10 @@ cat << EOF > input
-         if [[ $err -ne 0 ]]; then
-            fail=$(expr $fail + 1)
+         if [[ ${err} -ne 0 ]]; then
+            fail=$(expr ${fail} + 1)
```
```suggestion
            fail=$(( fail + 1 ))
```

WalterKolczynski-NOAA commented on ush/radmon_verf_bcor.sh:
```diff
@@ -126,18 +119,18 @@ else
          prep_step
 
-         ctr=$(expr $ctr + 1)
+         ctr=$(expr ${ctr} + 1)
```
```suggestion
         ctr=$(( ctr + 1 ))
```

WalterKolczynski-NOAA commented on ush/radmon_verf_bcor.sh:
```diff
@@ -126,18 +119,18 @@ else
-         if [[ $dtype == "anl" ]]; then
-            data_file=${type}_anl.${PDATE}.ieee_d
+         if [[ ${dtype} == "anl" ]]; then
+            data_file=${type}_anl.${PDY}${cyc}.ieee_d
```
```suggestion
            data_file="${type}_anl.${PDY}${cyc}.ieee_d"
```

WalterKolczynski-NOAA commented on ush/radmon_verf_bcor.sh:
```diff
@@ -126,18 +119,18 @@ else
             input_file=${type}_anl
          else
-            data_file=${type}.${PDATE}.ieee_d
+            data_file=${type}.${PDY}${cyc}.ieee_d
```
```suggestion
            data_file="${type}.${PDY}${cyc}.ieee_d"
```

WalterKolczynski-NOAA commented on ush/radmon_verf_bcor.sh:
```diff
@@ -151,7 +144,7 @@ else
       # the executable if $input_file doesn't exist or is 0 bytes
       #
-         if [[ -s $input_file ]]; then
+         if [[ -s ${input_file} ]]; then
```
```suggestion
         if [[ -s "${input_file}" ]]; then
```

WalterKolczynski-NOAA commented on ush/radmon_verf_bcor.sh:
```diff
@@ -173,10 +166,10 @@ cat << EOF > input
             export err=$?; err_chk
             if [[ $? -ne 0 ]]; then
-               fail=$(expr $fail + 1)
+               fail=$(expr ${fail} + 1)
```
```suggestion
               fail=$(( fail + 1 ))
```

WalterKolczynski-NOAA commented on ush/radmon_verf_bcor.sh:
```diff
@@ -185,36 +178,36 @@ EOF
 
    if compgen -G "bcor*.ieee_d*" > /dev/null || compgen -G "bcor*.ctl*" > /dev/null; then
-     tar -cf $tar_file bcor*.ieee_d* bcor*.ctl*
+     tar -cf ${tar_file} bcor*.ieee_d* bcor*.ctl*
```
```suggestion
     tar -cf "${tar_file}" bcor*.ieee_d* bcor*.ctl*
```

WalterKolczynski-NOAA commented on ush/radmon_verf_time.sh:
```diff
@@ -183,23 +170,23 @@ if [[ $err -eq 0 ]]; then
       fi
 
-      ctr=$(expr $ctr + 1)
+      ctr=$(expr ${ctr} + 1)
```
```suggestion
      ctr=$(( ctr + 1 ))
```

WalterKolczynski-NOAA commented on ush/radmon_verf_time.sh:
```diff
@@ -183,23 +170,23 @@ if [[ $err -eq 0 ]]; then
-         if [[ $dtype == "anl" ]]; then
-            data_file=${type}_anl.${PDATE}.ieee_d
+         if [[ ${dtype} == "anl" ]]; then
+            data_file=${type}_anl.${PDY}${cyc}.ieee_d
```
```suggestion
            data_file="${type}_anl.${PDY}${cyc}.ieee_d"
```

WalterKolczynski-NOAA commented on ush/radmon_verf_time.sh:
```diff
@@ -183,23 +170,23 @@ if [[ $err -eq 0 ]]; then
             time_ctl=time.${ctl_file}
          else
-            data_file=${type}.${PDATE}.ieee_d
+            data_file=${type}.${PDY}${cyc}.ieee_d
```
```suggestion
            data_file="${type}.${PDY}${cyc}.ieee_d"
```

WalterKolczynski-NOAA commented on ush/radmon_verf_time.sh:
```diff
@@ -227,48 +214,48 @@ cat << EOF > input
-         if [[ $err -ne 0 ]]; then
-            fail=$(expr $fail + 1)
+         if [[ ${err} -ne 0 ]]; then
+            fail=$(expr ${fail} + 1)
```
```suggestion
            fail=$(( fail + 1 ))
```

WalterKolczynski-NOAA commented on ush/radmon_verf_time.sh:
```diff
@@ -227,48 +214,48 @@ cat << EOF > input
    if compgen -G "time*.ieee_d*" > /dev/null || compgen -G "time*.ctl*" > /dev/null; then
      tar_file=radmon_time.tar
-     tar -cf $tar_file time*.ieee_d* time*.ctl*
+     tar -cf ${tar_file} time*.ieee_d* time*.ctl*
```
```suggestion
     tar -cf "${tar_file}" time*.ieee_d* time*.ctl*
```

WalterKolczynski-NOAA commented on ush/radmon_verf_time.sh:
```diff
@@ -344,12 +331,12 @@ EOF
-   qdate=$($NDATE -${CYCLE_INTERVAL} $PDATE)
-   pday=$(echo $qdate | cut -c1-8)
+   qdate=$(date --utc +%Y%m%d%H -d "${PDY} ${cyc} - ${assim_freq} hours")
+   pday=$(echo "${qdate}" | cut -c1-8)
```
```suggestion
   pday="${qdate:0:8}"
```

WalterKolczynski-NOAA reviewed (changes requested):
Could probably add 50 more places that need double-quotes just on the lines that were altered. Up to you if you want to go through them all or punt for now.

KateFriedman-NOAA committed:
Update qdate in ush/radmon_verf_time.sh

Refs #1925

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

KateFriedman-NOAA committed:
Apply suggestions from code review

Refs #1925

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/combine_vminmon_scripts

KateFriedman-NOAA committed:
Address shellcheck SC2126 in ush/radmon_diag_ck.sh

Refs #1925

KateFriedman-NOAA committed:
Address shellcheck SC2250 and SC2086

Update verfozn and verfrad driver scripts for shellcheck

Refs #1925

KateFriedman-NOAA commented:
> Could probably add 50 more places that need double-quotes just on the lines that were altered. Up to you if you want to go through them all or punt for now.

Made a few more updates in the driver scripts and accepted all of your suggestions. Thanks @WalterKolczynski-NOAA !

KateFriedman-NOAA commented:
Reran the monitor jobs in the test on WCOSS2 to make sure changes to address reviewer comments and further update based on shellcheck didn't break the jobs. Jobs succeeded.

aerorahul reviewed (approved):
Looks good.

KateFriedman-NOAA commented:
@EdwardSafford-NOAA Gentle poke to see if you'd like to review these changes before we merge this PR. Thanks!

EdwardSafford-NOAA reviewed (approved):
Looks good.  @KateFriedman-NOAA thanks for all your work cleaning these up.

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/combine_vminmon_scripts
